cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go

```diff
@@ -0,0 +1,249 @@
+package cluster
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	licorev1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/datawire/dlib/dlog"
+	"github.com/telepresenceio/telepresence/v2/pkg/subnet"
+)
+
+type nodeWatcher struct {
+	lister   licorev1.NodeLister
+	informer cache.SharedIndexInformer
+	subnets  []*net.IPNet
+	changed  time.Time
+	lock     sync.Mutex // Protects all access to ipsMap
+}
+
+func newNodeWatcher(lister licorev1.NodeLister, informer cache.SharedIndexInformer) *nodeWatcher {
+	return &nodeWatcher{
+		lister:   lister,
+		informer: informer,
+	}
+}
+
+func (w *nodeWatcher) changeNotifier(ctx context.Context, updateSubnets func([]*net.IPNet)) {
+	// Check for changes every 5 second
+	const nodeReviewPeriod = 5 * time.Second
+
+	// The time we wait from when the first change arrived until we actually do something. This
+	// so that more changes can arrive (hopefully all of them) before everything is recalculated.
+	const nodeCollectTime = 3 * time.Second
+
+	ticker := time.NewTicker(nodeReviewPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+		w.lock.Lock()
+		if w.changed.IsZero() || time.Since(w.changed) < nodeCollectTime {
+			w.lock.Unlock()
+			continue
+		}
+		w.changed = time.Time{}
+		subnets := make([]*net.IPNet, len(w.subnets))
+		copy(subnets, w.subnets)
+		w.lock.Unlock()
+		dlog.Debugf(ctx, "nodeWatcher calling updateSubnets with %v", subnets)
+		updateSubnets(subnets)
+	}
+}
+
+func (w *nodeWatcher) viable(ctx context.Context) bool {
+	nodes, err := w.lister.List(labels.Everything())
+	if err != nil {
+		dlog.Errorf(ctx, "unable to list nodes: %v", err)
+		return false
+	}
+
+	// Create the initial snapshot
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	changed := false
+	dlog.Infof(ctx, "Scanning %d nodes", len(nodes))
+	for _, node := range nodes {
+		if w.addLocked(nodeSubnets(ctx, node)) {
+			changed = true
+		}
+	}
+	if changed {
+		dlog.Infof(ctx, "Found %d subnets", len(w.subnets))
+		w.changed = time.Now()
+	} else {
+		dlog.Info(ctx, "No subnets found")
+	}
+	return changed
+}
```

Comment:
it seems like this function combines initialization with a precondition check (i.e. "Are we actually able to watch subnets from nodes?") -- that's okay, but calling it twice in a row will make the second call return `false`, no? Maybe we should short-circuit return if `w.subnets` is already populated -- after all, if they're there then getting them is definitely viable